### PR TITLE
Support Termination Requests #590

### PR DESCRIPTION
**Pull request by ap891843**
Support Termination Requests #590

LSP spec says below about "shutdown" call:
"Clients must not send any notifications other than exit or requests to a server to which they have sent a shutdown request. If a server receives requests after a shutdown request those requests should error with InvalidRequest."

While the VScode API doesn't mention anything specific about the call in [references/vscode-api#Extension](https://code.visualstudio.com/api/references/vscode-api#Extension). It does mention it in [references/activation-events](https://code.visualstudio.com/api/references/activation-events)

And an issue [vscode issue](https://github.com/microsoft/vscode/issues/35196), mentions "VScode only wait for shutdown (e.g. for the promise to be resolved) and wait in total 5 seconds. There is not waiting happening on exit."

### CODE REVIEW

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java [hunk 1]
```diff
@@ -124,14 +125,21 @@ private void getLocaleFromClient() {
     settingsService.getConfiguration(LOCALE.label).thenAccept(localeStore.notifyLocaleStore());
   }
 
+  public int getExitCode() {
+    return exitCode;
+  }
+
   @Override
   public CompletableFuture<Object> shutdown() {
+    LOG.info("LS for COBOL received shutdown request");
```
Comment (ap891843) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java:134]: As per LSP spec, after this call no other service call can/should be handled. Should we have a logic to return no-valid from all our service after this call is made. 
Why I didn't implement it? 
VScode actually never waits for us to call exit. Upon closing, it fires a shutdown call and then fires an exit call.
Details mentioned in the PR comment.

Comment (temanbrcom) [reply server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java:134]: We support different IDEs, should it would be better not to rely on the client implementation

File: clients/cobol-lsp-vscode-extension/src/__tests__/LanguageClientServiceTest.spec.ts [hunk 2]
```diff
@@ -104,4 +104,12 @@ describe("LanguageClientService positive scenario", () => {
             },
         });
     });
+
+    test("Test LanguageClientService exit language server is calls exit on languageClient", () => {
```
Comment (temanbrcom) [inline clients/cobol-lsp-vscode-extension/src/__tests__/LanguageClientServiceTest.spec.ts:108]: It is not clear to me what does this test checks from the description

File: clients/cobol-lsp-vscode-extension/src/services/LanguageClientService.ts [hunk 3]
```diff
@@ -43,11 +43,19 @@ export class LanguageClientService {
...
+        const languageClient = this.getLanguageClient();
+        return languageClient.start();
+    }
+
+    public stop(): Thenable<void> {
+        return this.getLanguageClient().stop();
+    }
+
+    private getLanguageClient() {
+        return new LanguageClient(LANGUAGE_ID,
```
Comment (temanbrcom) [inline clients/cobol-lsp-vscode-extension/src/services/LanguageClientService.ts:55]: Doesn't it mean that it will stop a new `LanguageClient` instance?

File: server/src/main/java/com/broadcom/lsp/cobol/core/annotation/CheckServerShutdownState.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.broadcom.lsp.cobol.core.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface CheckServerShutdownState {}
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/core/annotation/CheckServerShutdownState.java:10]: Please, add a Javadoc

File: server/src/main/java/com/broadcom/lsp/cobol/core/annotation/CheckServerShutdownState.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.broadcom.lsp.cobol.core.annotation;
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/core/annotation/CheckServerShutdownState.java:1]: Missing copyright header

File: server/src/main/java/com/broadcom/lsp/cobol/core/annotation/HandleShutdownState.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package com.broadcom.lsp.cobol.core.annotation;
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/core/annotation/HandleShutdownState.java:1]: Missing copyright header

File: server/src/main/java/com/broadcom/lsp/cobol/core/annotation/HandleShutdownState.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import com.google.inject.Inject;
+import lombok.extern.slf4j.Slf4j;
+import org.aopalliance.intercept.MethodInterceptor;
+import org.aopalliance.intercept.MethodInvocation;
+import org.eclipse.lsp4j.services.LanguageServer;
+
+import java.util.concurrent.CompletableFuture;
+
+@Slf4j
+public class HandleShutdownState implements MethodInterceptor {
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/core/annotation/HandleShutdownState.java:13]: Missing Javadoc

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java [hunk 8]
```diff
@@ -213,11 +224,13 @@ private void interruptAnalysis(String uri) {
   }
 
   @Override
+  @CheckServerShutdownState
   public void didSave(DidSaveTextDocumentParams params) {
     LOG.info("Document saved...");
   }
 
   @Override
+  @CheckServerShutdownState
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java:233]: Not an LSP request, no need to check it

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java [hunk 9]
```diff
@@ -145,4 +187,11 @@ private ExecuteCommandOptions collectExecuteCommandList() {
     return new ExecuteCommandOptions(
         stream(ErrorCode.values()).map(ErrorCode::name).collect(toList()));
   }
+
+  /** Represents the JSON RPC response structure for shutdown command as per LSP specification */
+  @AllArgsConstructor
+  private class ShutdownResponse {
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java:193]: Nice idea to create a class for it, but it should be either static or a separate class because it is not strongly connected with the outer class.

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java [hunk 10]
```diff
@@ -124,14 +127,53 @@ private void getLocaleFromClient() {
...
+   *
+   * <p>Source:
+   * https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ThreadPoolExecutor.html#shutdownNow--
+   */
+  private void cancelAllProcessing() {
+    // NOTE: shutdownNow principally is not a bad practice when we know what we are doing.
+    //      Here in this case, since we are only stopping our own threads which don't
+    //      require any clean up.It's ok. Java doc doesn't mention any bad practice for this.
+
+    // cancels all the running task and clears empty queue ,initiated on forkJoinPool.
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java:164]: A misplaced comma in the comment, as well as in a comment below

File: server/src/main/java/com/broadcom/lsp/cobol/core/strategy/Messages.java [hunk 11]
```diff
@@ -1,51 +0,0 @@
-/*
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/core/strategy/Messages.java:1]: Why did you remove this class?

File: server/src/test/java/com/broadcom/lsp/cobol/service/CobolTextDocumentServiceTest.java [hunk 12]
```diff
@@ -437,6 +441,58 @@ void testImmediateClosingOfDocumentDoNotCauseNPE() {
...
+   * @throws ExecutionException
+   * @throws InterruptedException
+   */
+  @Test
+  void whenShutdownIsFired_ThenNewRequestReturnInvalidResponse()
+      throws ExecutionException, InterruptedException {
+    CobolTextDocumentService service =
+        (CobolTextDocumentService)
+            LangServerCtx.getInjector().getInstance(TextDocumentService.class);
+    LanguageServer server = LangServerCtx.getInjector().getInstance(LanguageServer.class);
```
Comment (temanbrcom) [inline server/src/test/java/com/broadcom/lsp/cobol/service/CobolTextDocumentServiceTest.java:457]: Please, don't use this type of DI in tests

**Review by temanbrcom — CHANGES_REQUESTED** [review]

**Comment by temanbrcom** [comment]
@asatklichov 
@grianbrcom 
@Nurkambay

File: clients/cobol-lsp-vscode-extension/src/services/LanguageClientService.ts (hunk 3 above)
Comment (ap891843) [reply clients/cobol-lsp-vscode-extension/src/services/LanguageClientService.ts:55]: yes, this is the wrong implementation.

File: server/src/main/java/com/broadcom/lsp/cobol/core/strategy/Messages.java (hunk 11 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/core/strategy/Messages.java:1]: I guess we don't use this class now and all the messages are being handled through the resource bundle. I will check this again. Did see this yesterday in the Coverity Scan, so thought of removing it.

**Comment by ap891843** [comment]
Converting this PR to draft, as currently struggling with some test cases.

File: server/src/main/java/com/broadcom/lsp/cobol/core/strategy/Messages.java (hunk 11 above)
Comment (temanbrcom) [reply server/src/main/java/com/broadcom/lsp/cobol/core/strategy/Messages.java:1]: I agree to delete it, but please do it as a separate commit not to bring a mess to this one

File: clients/cobol-lsp-vscode-extension/src/__tests__/LanguageClientServiceTest.spec.ts [hunk 13]
```diff
@@ -104,4 +104,12 @@ describe("LanguageClientService positive scenario", () => {
...
+        jest.clearAllMocks();
+        new JavaCheck().isJavaInstalled = jest.fn().mockResolvedValue(true);
+        vscode.workspace.getConfiguration(expect.any(String)).get = jest.fn().mockReturnValue(9999);
```
Comment (grianbrcom) [inline clients/cobol-lsp-vscode-extension/src/__tests__/LanguageClientServiceTest.spec.ts:111]: The test pass even if I remove those three lines. I suggest to remove them.

File: clients/cobol-lsp-vscode-extension/src/extension.ts [hunk 14]
```diff
@@ -32,18 +32,18 @@ import {TelemetryService} from "./services/reporter/TelemetryService";
 import {createFileWithGivenPath, initializeSettings} from "./services/Settings";
 import {ZoweApi} from "./services/ZoweApi";
 
+const zoweApi: ZoweApi = new ZoweApi();
+const profileService: ProfileService = new ProfileService(zoweApi);
+const copybooksPathGenerator: CopybooksPathGenerator = new CopybooksPathGenerator(profileService);
+const copyBooksDownloader: CopybookDownloadService =
+new CopybookDownloadService(zoweApi, profileService, copybooksPathGenerator);
```
Comment (grianbrcom) [inline clients/cobol-lsp-vscode-extension/src/extension.ts:39]: Could you please keep in on one line? Or add indentation on 39 line.

File: server/src/main/java/com/broadcom/lsp/cobol/core/annotation/HandleShutdownState.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    this.server = server;
+  }
+
+  @Override
+  public Object invoke(MethodInvocation invocation) throws Throwable {
+    if (server instanceof DisposableLanguageServer) {
+      DisposableLanguageServer languageServer = (DisposableLanguageServer) server;
+      if (languageServer.getExitCode() == DisposableLanguageServer.SHUTDOWN_EXIT_CODE) {
+        LOG.info(invocation.getMethod().getName() + " invoked after shutdown");
+        return CompletableFuture.supplyAsync(() -> SHUTDOWN_RESPONSE);
```
Comment (grianbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/core/annotation/HandleShutdownState.java:28]: You don't really need to resolve this future asynchronously. Please, use this instead:
```
return CompletableFuture.completedFuture(SHUTDOWN_RESPONSE);
```

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java [hunk 16]
```diff
@@ -124,14 +127,53 @@ private void getLocaleFromClient() {
...
+
   @Override
   public CompletableFuture<Object> shutdown() {
-    return supplyAsync(() -> TRUE);
+    LOG.info("LS for COBOL received shutdown request");
+    exitCode = 0;
+    try {
+      cancelAllProcessing();
+    } catch (Exception exception) {
+      return supplyAsync(() -> new ShutdownResponse(null, exception.getMessage()));
```
Comment (grianbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java:142]: You can use `CompletableFuture.completedFuture` in both returns.

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java [hunk 17]
```diff
@@ -124,14 +127,53 @@ private void getLocaleFromClient() {
...
+  @Override
+  public int getExitCode() {
+    return exitCode;
+  }
+
   @Override
   public CompletableFuture<Object> shutdown() {
-    return supplyAsync(() -> TRUE);
+    LOG.info("LS for COBOL received shutdown request");
+    exitCode = 0;
```
Comment (grianbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java:138]: It makes sense to update exitCode after successfully done the `cancelAllProcessing` call.

File: clients/cobol-lsp-vscode-extension/src/__tests__/LanguageClientServiceTest.spec.ts (hunk 2 above)
Comment (ap891843) [reply clients/cobol-lsp-vscode-extension/src/__tests__/LanguageClientServiceTest.spec.ts]: updated

File: clients/cobol-lsp-vscode-extension/src/services/LanguageClientService.ts (hunk 3 above)
Comment (ap891843) [reply clients/cobol-lsp-vscode-extension/src/services/LanguageClientService.ts]: updated

File: server/src/main/java/com/broadcom/lsp/cobol/core/annotation/CheckServerShutdownState.java (hunk 5 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/core/annotation/CheckServerShutdownState.java:16]: done

File: server/src/main/java/com/broadcom/lsp/cobol/core/annotation/HandleShutdownState.java (hunk 6 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/core/annotation/HandleShutdownState.java:16]: done

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java (hunk 8 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java:234]: I spent major time here. Using interceptor actually leads to some proxy classes and it seems without this annotation here, this method is marked final which leads to error while spying. I get the below error

> java.lang.VerifyError: class com.broadcom.lsp.cobol.service.CobolTextDocumentService$$EnhancerByGuice$$c86ee31e$MockitoMock$565924183 overrides final method com.broadcom.lsp.cobol.service.CobolTextDocumentService$$EnhancerByGuice$$c86ee31e.observerCallback(Ljava/lang/Object;)

This confuses me if guice interceptors are the right approach.  This is a question to all reviewers, if you feel the same , I will re-work this with some other approach.
For now, I have kept this annotation as is.

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java (hunk 9 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java]: done

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java (hunk 10 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java]: done

File: server/src/main/java/com/broadcom/lsp/cobol/core/strategy/Messages.java (hunk 11 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/core/strategy/Messages.java]: done

File: server/src/test/java/com/broadcom/lsp/cobol/service/CobolTextDocumentServiceTest.java (hunk 12 above)
Comment (ap891843) [reply server/src/test/java/com/broadcom/lsp/cobol/service/CobolTextDocumentServiceTest.java:457]: done as per the rest of the test class

File: clients/cobol-lsp-vscode-extension/src/__tests__/LanguageClientServiceTest.spec.ts (hunk 13 above)
Comment (ap891843) [reply clients/cobol-lsp-vscode-extension/src/__tests__/LanguageClientServiceTest.spec.ts]: done

File: clients/cobol-lsp-vscode-extension/src/extension.ts (hunk 14 above)
Comment (ap891843) [reply clients/cobol-lsp-vscode-extension/src/extension.ts]: done

File: server/src/main/java/com/broadcom/lsp/cobol/core/annotation/HandleShutdownState.java (hunk 15 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/core/annotation/HandleShutdownState.java]: done

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java (hunk 16 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java]: done

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java (hunk 17 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java]: done

File: server/src/test/java/com/broadcom/lsp/cobol/service/CobolTextDocumentServiceTest.java (hunk 12 above)
Comment (ap891843) [reply server/src/test/java/com/broadcom/lsp/cobol/service/CobolTextDocumentServiceTest.java:457]: working on it, seems like I overlooked

Comment (ap891843) [reply server/src/test/java/com/broadcom/lsp/cobol/service/CobolTextDocumentServiceTest.java]: done

**Review by temanbrcom — APPROVED** [review]
Looks cool!

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java [hunk 18]
```diff
@@ -126,12 +134,45 @@ private void getLocaleFromClient() {
...
+
+    // cancels all the running task and clears empty queue on CobolTextDocumentService executor.
+    ((CobolTextDocumentService) getTextDocumentService()).disposeAllOperation(this);
+    LOG.info("All processing abandoned as per shutdown call");
   }
 
   @Override
   public void exit() {
-    // not supported
+    LOG.info("LS for COBOL exiting with code: " + exitCode);
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java:174]: I just tested it and noticed that this message should be "COBOL LS server exits with code:"

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java [hunk 19]
```diff
@@ -126,12 +134,45 @@ private void getLocaleFromClient() {
 
   @Override
   public CompletableFuture<Object> shutdown() {
-    return supplyAsync(() -> TRUE);
+    LOG.info("LS for COBOL received shutdown request");
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java:137]: Should be "COBOL LS received shutdown request"

Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java]: done

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java (hunk 18 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java]: Done

File: clients/cobol-lsp-vscode-extension/src/services/LanguageClientService.ts [hunk 20]
```diff
@@ -43,11 +44,23 @@ export class LanguageClientService {
     }
 
     public start(): vscode.Disposable {
-        const languageClient = new LanguageClient(LANGUAGE_ID,
-            "LSP extension for " + LANGUAGE_ID + " language",
-            this.createServerOptions(this.jarPath),
-            this.createClientOptions());
-        return languageClient.start();
+        return this.getLanguageClient().start();
```
Comment (ap891843) [inline clients/cobol-lsp-vscode-extension/src/services/LanguageClientService.ts:47]: Reverting back custom thread pool changes for this class.

Reason:
Manage thread for the process we initiate and let LSP4J manage its thread. This helps in shutdown process where we shutdown the custom thread pool to shutdown all current processes.

File: server/src/main/java/com/broadcom/lsp/cobol/LangServerBootstrap.java [hunk 21]
```diff
@@ -56,23 +54,19 @@ public static void main(String[] args)
     Injector injector = initCtx();
     LanguageServer server = injector.getInstance(LanguageServer.class);
     ClientProvider provider = injector.getInstance(ClientProvider.class);
-    CustomThreadPoolExecutor customExecutor = injector.getInstance(CustomThreadPoolExecutor.class);
 
-    start(args, server, provider, customExecutor.getThreadPoolExecutor());
+    start(args, server, provider);
```
[the remark this answers is not in the document]
Comment (temanbrcom) [reply server/src/main/java/com/broadcom/lsp/cobol/LangServerBootstrap.java:58]: But in this case, we still have a separate executor for LSP4J and another one for the data bus and other stuff, right?

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java [hunk 22]
```diff
@@ -126,12 +138,45 @@ private void getLocaleFromClient() {
...
+  private void cancelAllProcessing() {
+    // NOTE: shutdownNow principally is not a bad practice when we know what we are doing.
+    //      Here in this case, since we are only stopping our own threads which don't
+    //      require any clean up.It's ok. Java doc doesn't mention any bad practice for this.
+
+    // cancels all the running task and clears empty queue on forkJoinPool.
+    ForkJoinPool.commonPool().shutdownNow();
+
+    // cancels all the running task on COBOL LS custom executor service.
+    customThreadPoolExecutor.getThreadPoolExecutor().shutdownNow();
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java:172]: Should we shutdown also the scheduled pool here?

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java [hunk 23]
```diff
@@ -87,14 +92,16 @@
       Completions completions,
       Occurrences occurrences,
       DataBusBroker dataBus,
-      CodeActions actions) {
+      CodeActions actions,
+      CustomThreadPoolExecutor customThreadPoolExecutor) {
```
Comment (temanbrcom) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java:96]: Optional: this name seems very wordy, especially on the usages. Maybe, `executors` would be enough?

File: server/src/main/java/com/broadcom/lsp/cobol/LangServerBootstrap.java (hunk 21 above)
[the remark this answers is not in the document]
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/LangServerBootstrap.java:58]: Yes, LSP4J uses a new cached thread pool for its use, if we don't provide (Executors.newCachedThreadPool).
And our process will use the custom thread pool (dataBus/ textDocumnetService etc).

File: server/src/test/java/com/broadcom/lsp/cobol/service/CobolTextDocumentServiceTest.java [hunk 24]
```diff
@@ -178,7 +183,14 @@ void disableCopybookAnalysisOnExtendedDoc() {
...
     return new CobolTextDocumentService(
-        mock(Communications.class), engine, null, null, null, mock(DataBusBroker.class), null);
+        mock(Communications.class),
+        engine,
+        null,
+        null,
+        null,
+        mock(DataBusBroker.class),
+        null,
+        getCustomExecutor());
```
Comment (temanbrcom) [inline server/src/test/java/com/broadcom/lsp/cobol/service/CobolTextDocumentServiceTest.java:193]: Replaceable with builder

Comment (ap891843) [reply server/src/test/java/com/broadcom/lsp/cobol/service/CobolTextDocumentServiceTest.java]: done

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java (hunk 23 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java]: done

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java (hunk 22 above)
Comment (ap891843) [reply server/src/main/java/com/broadcom/lsp/cobol/service/CobolLanguageServer.java:172]: done

**Review by grianbrcom — APPROVED** [review]
:flying_disc:

File: server/src/test/java/com/broadcom/lsp/cobol/service/DocumentExtensionTests.java [hunk 25]
```diff
@@ -34,7 +35,7 @@
  * extensions. If an extension of a document is unsupported (i.e. not "cob", "cbl" or "cobol"
  * ignoring case) then the according to notification should be sent.
  */
-class DocumentExtensionTests {
+class DocumentExtensionTests extends ConfigurableTest {
```
Comment (temanbrcom) [inline server/src/test/java/com/broadcom/lsp/cobol/service/DocumentExtensionTests.java:38]: In this case, we will create a connection to DI that we are actually want to get rid of. Will work for now; may be fixed in #261

**Review by temanbrcom — APPROVED** [review]
Looks ok for me, but I would like @asatklichov to also check it

File: server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java [hunk 26]
```diff
@@ -108,70 +115,85 @@
...
 
   @Override
+  @CheckServerShutdownState
   public CompletableFuture<Either<List<CompletionItem>, CompletionList>> completion(
       CompletionParams params) {
     String uri = params.getTextDocument().getUri();
     return CompletableFuture.<Either<List<CompletionItem>, CompletionList>>supplyAsync(
-            () -> Either.forRight(completions.collectFor(docs.get(uri), params)))
+            () -> Either.forRight(completions.collectFor(docs.get(uri), params)),
+            executors.getThreadPoolExecutor())
```
Comment (asatklichov) [inline server/src/main/java/com/broadcom/lsp/cobol/service/CobolTextDocumentService.java:124]: 👍

**Review by asatklichov — APPROVED** [review]
🍦
